[MRG+2] Deprecate n_iter in SGDClassifier and implement max_iter

TomDLT opened the pull request:
Solve #5022
In SGDClassifier, SGDRegressor, Perceptron, PassiveAgressive:
- **Deprecate `n_iter`.** Default is now `None`. If not None, it warns and sets `max_iter = n_iter` and `tol = 0`, to have exact previous behavior.
- **Implement `max_iter` and `tol`.** The stopping criterion in `sgd_fast._plain_sgd()` is identical to the one in SAG new solver for Ridge and LogisticRegression.
- **Add `self.n_iter_` after the fit.** For multiclass classifiers, we keep the maximum `n_iter_` over all binary (OvA) fits.

ogrisel commented on sklearn/linear_model/sgd_fast.pyx:
```diff
@@ -700,7 +728,7 @@ def _plain_sgd(np.ndarray[double, ndim=1, mode='c'] weights,
 
     w.reset_wscale()
 
```
Please raise a `ConvergenceWarning` with an informative message if `max_iter == epoch + 1`.

TomDLT replied:
It would raise a warning for each partial_fit (which has `max_iter=1`).
Instead I suggest to raise it in the [_fit method](https://github.com/scikit-learn/scikit-learn/pull/5036/files#diff-c58bb22c5737f200d2494695f6dea33dR451).

ogrisel replied:
You could disable the convergence warning by passing `tol=0` as a local variable only when called from `partial_fit` while passing `tol=self.tol` when called from `fit`.

ogrisel commented:
About your remark in #5022, you suggested that we could avoid deprecating `n_iter` and instead leave it to `None` by default so that the `max_iter` + `tol` criterion would be used by default (issue a `ConvergenceWarning` if `tol` is not reached before `max_iter`) while still leaving the user the ability to ask for a specific `n_iter` passes over the data in which case `max_iter` and `tol` are ignored and `ConvergenceWarning` is never raised.

I am +0 for this convenience feature. @amueller do you have an opinion in this regard?

amueller commented:
I'm -0 ;)

Do we want a validation set for SGDClassifier in the future? And if so, do we introduce this using a ``validation=boolean`` variable? This is a step in the right direction but we should think about where this is going.

TomDLT commented:
>we could avoid deprecating `n_iter`

I am not sure to understand what would be the classic deprecation.

> Do we want a validation set for SGDClassifier in the future?

Do you mean to give a validation set with a performance goal as a stopping criterion, directly in the SGD solver?

ogrisel commented:
> I though it as deprecation, a temporary way not to break any user code, before removing completely n_iter.
I am not sure to understand what would be the classic deprecation.

Classic deprecation is what you did in this PR: raise a DeprecationWarning now while still behaving the same if the user is passing a `n_iter != None` and remove the `n_iter` kwarg completely in 2 releases.

>> Do we want a validation set for SGDClassifier in the future?...

> Do you mean to give a validation set with a performance goal as a stopping criterion, directly in the SGD solver?

Yes early stopping on the lack of improvement as measured on a validation set. The validation set can be specified as number between 0 and 1 (typically 0.1 by default) and the model extracts internally in the fit method by randomly splitting the user provided data into train and validation folds. But this is outside of the scope of this PR.

ogrisel commented on sklearn/linear_model/passive_aggressive.py:
```diff
@@ -81,20 +89,24 @@ class PassiveAggressiveClassifier(BaseSGDClassifier):
-                 verbose=0, loss="hinge", n_jobs=1, random_state=None,
-                 warm_start=False, class_weight=None):
+
+    def __init__(self, C=1.0, fit_intercept=True, max_iter=5, tol=1e-4,
```
Now that we have a good stopping criterion, I think we should set `max_iter=100` by default and expect the stopping criterion to kick in  before that in 99% of the cases.

amueller replied:
On 09/08/2015 02:33 AM, Olivier Grisel wrote:
>
> Now that we have a good stopping criterion, I think we should set 
> |max_iter=100| by default and expect the stopping criterion to kick in 
> before that in 99% of the cases.
>
Did someone do experiments on how well that works in practice?

ogrisel commented on sklearn/linear_model/passive_aggressive.py:
```diff
@@ -183,10 +195,14 @@ class PassiveAggressiveRegressor(BaseSGDRegressor):
-        The number of passes over the training data (aka epochs).
+    max_iter : int, optional
+        The maximum number of passes over the training data (aka epochs).
+        The maximum number of iterations is set to 1 if using partial_fit.
```
I would rather say that this parameter only impacts the behavior of the `fit` method, not the `partial_fit` method.

ogrisel commented on sklearn/linear_model/passive_aggressive.py:
```diff
@@ -228,20 +247,23 @@ class PassiveAggressiveRegressor(BaseSGDRegressor):
-    def __init__(self, C=1.0, fit_intercept=True, n_iter=5, shuffle=True,
-                 verbose=0, loss="epsilon_insensitive",
-                 epsilon=DEFAULT_EPSILON, random_state=None, warm_start=False):
+    def __init__(self, C=1.0, fit_intercept=True, max_iter=5, tol=1e-4,
```
`max_iter=100` as well here.

ogrisel commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -73,6 +76,13 @@ def __init__(self, loss, penalty='l2', alpha=0.0001, C=1.0,
         self.average = average
 
+        if n_iter is not None:
+            warnings.warn("n_iter parameter is deprecated and will be removed"
```
It's better to be very explicit: `"n_iter parameter is deprecated in 0.17 and will be removed in 0.19. ..."`

ogrisel commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -426,21 +441,27 @@ def _fit(self, X, y, alpha, C, loss, learning_rate, coef_init=None,
+        self._partial_fit(X, y, alpha, C, loss, learning_rate, self.max_iter,
                           classes, sample_weight, coef_init, intercept_init)
 
+        if self.n_iter_ == self.max_iter:
```
Please change this test to:

```python
if self.tol > 0 and self.n_iter_ == self.max_iter:
```

so that the user can disable the `ConvergenceWarning` when he/she decide to always perform `max_iter` iterations intentionally (effectively disabling the stopping criterion).

ogrisel commented:
The handling of `partial_fit` is wrong. The `n_iter_` attribute should not necessarily be defined when the users calls only `partial_fit` to train the model as the model is not aw

```python
>>> from sklearn.datasets import load_boston
>>> from sklearn.linear_model import SGDRegressor
>>> from sklearn.utils import gen_batches
>>> boston = load_boston()
>>> n_samples, n_features = boston.data.shape
>>> n_samples, n_features
(506, 13)
>>> all_batches = list(gen_batches(n_samples, 100))
>>> m = SGDRegressor(max_iter=2)
>>> for batch in all_batches:
...     m.fit(boston.data[batch], boston.target[batch])
...
>>> m.t_
13.0
```

`m.t_` should be the total number of samples has seen since the beginning of the incremental fit. Iterating over `all_batches` should be equivalent to performing one pass over the data (one epoch) and therefore `m.t_` should be `n_samples + 1 = 507` instead of currently: `max_iter * len(all_batches) + 1`.

In particular, calling

```python
m = SGDRegressor(max_iter=1, tol=0, shuffle=False, random_state=0)
m.fit(boston.data, boston.target)
```

should be equivalent (same `m.t_`, `m.coef_` and `m.intercept_`) to:

```python
# max_iter should not impact incremental fitting at all
m = SGDRegressor(max_iter=42, shuffle=False, random_state=0)
for batch in all_batches:
    m.partial_fit(boston.data[batch], boston.target[batch])
```

`m.t_` should be `n_samples + 1` in both cases.

Furthermore:

```python
m = SGDRegressor(max_iter=10, tol=0, shuffle=False, random_state=0)
m.fit(boston.data, boston.target)
```

should be equivalent to:

```python
m = SGDRegressor(max_iter=42, shuffle=False, random_state=0)
for i in 10:
    for batch in all_batches:
        m.partial_fit(boston.data[batch], boston.target[batch])
```

`m.t_` should be `10 * n_samples + 1` in both cases.

The fact that the tests do not fail mean that `partial_fit` testing is not strong enough for `SGDRegressor` (and `SGDClassifier` and related models).

ogrisel commented:
Also for some reason I don't get the expected convergence warning when I do:

```python
SGDRegressor(max_iter=1, tol=1e-15).fit(boston.data, boston.target)
```

ogrisel commented:
On the other hand, I would expect the following model to converge earlier (`n_iter_` should be smaller than `max_iter`):

```python
>>> SGDRegressor(max_iter=10000, tol=1e-2).fit(boston.data, boston.target).n_iter_
10000
```

TomDLT commented:
Thanks for the review.

About your last comment, it comes from the fact that the SGD converges quite slowly with the boston dataset, as you can see in the following plot (tested on master):
```Python
import numpy as np
import matplotlib.pyplot as plt

from sklearn.datasets import load_boston
from sklearn.linear_model import SGDRegressor

boston = load_boston()
X, y = boston.data, boston.target
n_features = X.shape[1]

iter_range = np.arange(1, 11) * 1000
coefs = np.zeros((n_features, iter_range.size))

for i, n_iter in enumerate(iter_range):
    reg = SGDRegressor(n_iter=n_iter).fit(X, y)
    coefs[:, i] = reg.coef_

for i in range(n_features):
    plt.plot(iter_range, coefs[i, :])

plt.xlabel("n_iter")
plt.ylabel("coefs")
plt.show()
```
![unstable_boston](https://cloud.githubusercontent.com/assets/11065596/9730886/d2b48466-5619-11e5-9701-69c4fbef5b42.png)

___
I changed the default `max_iter`, I added the `ConvergenceWarning` in `SGDRegressor`, and I am looking into your comment about `t_` and `partial_fit`.

TomDLT commented:
Actually I think the code is OK about `t_`.

```python
>>>all_batches
[slice(0, 100, None),
 slice(100, 200, None),
 slice(200, 300, None),
 slice(300, 400, None),
 slice(400, 500, None),
 slice(500, 506, None)]
>>> for batch in all_batches:
...     m.fit(boston.data[batch], boston.target[batch])
...
>>> m.t_
13.0
```
You call `fit` on the last slice (with only 6 samples), with `max_iter=2`, so it is normal that `m.t_= 2 * 6 +1`. Warm starting is not activated by default, and I guess we want `t_` to restart even in the case of warm starting.

With `partial_fit`, `t_` seems correct :
```python
from sklearn.datasets import load_boston
from sklearn.linear_model import SGDRegressor
from sklearn.utils import gen_batches

boston = load_boston()
n_samples, n_features = boston.data.shape
all_batches = list(gen_batches(n_samples, 100))

for max_iter in range(1, 11):
    # one full pass with fit
    m1 = SGDRegressor(max_iter=max_iter, tol=0, shuffle=False, random_state=0)
    m1.fit(boston.data, boston.target)

    # batches with partial_fit
    m2 = SGDRegressor(max_iter=42, shuffle=False, random_state=0)
    for _ in range(max_iter):
        for batch in all_batches:
            m2.partial_fit(boston.data[batch], boston.target[batch])

    print("%d, %f" % (m1.t_, m1.coef_[1]))
    print("%d, %f" % (m2.t_, m2.coef_[1]))
```
gives
```
507, 216965230251.886780
507, 216965230251.886414
1013, 891040660054.554077
1013, 891040660054.553955
1519, 724458977715.773315
1519, 724458977715.773071
2025, 753346099309.831177
2025, 753346099309.830688
2531, 836014190413.200317
2531, 836014190413.200317
3037, 942892124924.625488
3037, 764967807987.214600
3543, 365984307776.865479
3543, 498725629682.746338
4049, 81118161082.128448
4049, 575624185079.631470
4555, 47289103200.144722
4555, 618775289008.635620
5061, -32564036946.908470
5061, 266499003140.677155
```
`partial_fit` and `fit` give almost the same results. I don't understand why they are not perfectly identical. SGD is very unstable on Boston dataset (cf. previous plot), which explains the large difference with more iterations. This behavior was already present in master branch.

About `n_iter_` with `partial_fit`, is it really misleading to let it at 1?

amueller commented:
for default parameters of SGD to work, the data needs to be scaled.
Maybe we should warn / add this to SGDClassifier and SGDRegressor?

amueller commented:
Tag 0.17 @ogrisel ? Not sure if it's ready.

amueller commented:
I'm not sure if we should do this change simultaneously with a default scaling change?

TomDLT commented:
> for default parameters of SGD to work, the data needs to be scaled.

Indeed, adding a `StandardScaler` step solves the unstable behavior in previous plot. And then `fit` and `partial_fit` do have the exact same values. Thanks

> I'm not sure if we should do this change simultaneously with a default scaling change?

The problem of scaling is rather not linked to `n_iter` and `max_iter`.
I would suggest to make a different PR.
___
I think this PR is OK, except if I missed your point @ogrisel

amueller commented:
Hm this is breaking behavior, right?
Shouldn't we add ``tol=None`` and warn that it'll change in the future?

TomDLT commented:
`SGD(n_iter=n_iter)`'s behavior is not changed and raise DeprecationWarning.
`SGD()`'s behavior is changed however.
It might be better to warn a future change.
But if we add `tol=None` we need to change back to `max_iter=5`.

amueller commented:
True. I'm just a bit concerned that SGDClassifier() will now behave wildly different. It will probably take much longer on most problems, right? And in general behavior change is bad.... Usually we'd keep the old default behavior for two versions...

TomDLT commented:
**update:**
I changed the default values not to modify any previous behaviors:

Default values are now `max_iter=5, tol=None, n_iter=None`.
 - If `n_iter` is specified, a DeprecationWarning is raised to warn about future removal, but the behavior is unchanged, using `self.max_iter=n_iter` and `self.tol=0.`.
 - If `tol` is not specified, a DeprecationWarning is raised to warn about future change in default value, but the behavior is not changed (`tol=None` behaves the same as `tol=0.`).

ogrisel commented:
I remove the 0.17 milestone because I won't have the time to properly review this before the 0.17.X branch cut and would rather like to focus on bug fixing for the release. Will review in depth soon-ish, I promise ;)

amueller commented:
Is this for 0.18? ping @ogrisel ?

amueller commented:
thanks @TomDLT for the 0.19 change, sorry we didn't get this in. I'll try to review. Or maybe @jnothman  ;)

jnothman commented on sklearn/utils/weight_vector.pyx:
```diff
@@ -197,3 +203,28 @@ cdef class WeightVector(object):
+            max_weight = fmax(max_weight, fabs(w_data_ptr[j]))
+        return max_weight
+
+    cdef double get_max_change(self, double **previous_weights) nogil:
```
It would be nice if this could be reused in sag, but it's no big deal...

TomDLT replied:
I don't remember why we didn't use `weight_vectors` for SAG in the first place, but it will need a bit of work to update it.

jnothman reviewed (changes requested):
I don't see any validation that `tol` is working (large `tol` should yield smaller `n_iter_` than small `tol`), that it validates correctly, or that the warnings are raised appropriately.

TomDLT commented:
Thanks for the review, I updated accordingly.

jnothman commented on sklearn/linear_model/tests/test_sgd.py:
```diff
@@ -1177,6 +1180,54 @@ def test_large_regularization():
+    model_2 = SGDClassifier(tol=1e-1, max_iter=max_iter).fit(X, y)
+    assert_greater(model_1.n_iter_, model_2.n_iter_)
+
+    # A larger tol should yield a smaller number of iteration
```
This may be overkill. Is the test quick?

TomDLT replied:
This tests lasts 50 ms, removing `model_3` does not speed it up since there is few iterations compared to `model_0` or `model_1`.

I changed `model_0` to go down to 20 ms.

The other tests lasts 9 ms.

agramfort commented:
+1 for MRG

GaelVaroquaux commented on benchmarks/bench_sgd_regression.py:
```diff
@@ -70,30 +70,30 @@
-                               n_iter=n_iter, learning_rate="invscaling",
-                               eta0=.01, power_t=0.25)
+                               max_iter=max_iter, learning_rate="invscaling",
+                               eta0=.01, power_t=0.25, tol=1e-12)
```
The tol is there to make sure that the iteration goes all the way to the end?

If so, it might be a good idea to add a comment.

GaelVaroquaux commented on benchmarks/bench_covertype.py:
```diff
@@ -102,7 +102,7 @@ def load_data(dtype=np.float32, order='C', random_state=13):
     'RandomForest': RandomForestClassifier(n_estimators=20),
     'CART': DecisionTreeClassifier(min_samples_split=5),
-    'SGD': SGDClassifier(alpha=0.001, n_iter=2),
+    'SGD': SGDClassifier(alpha=0.001, max_iter=2, tol=0),
```
The tol is there to make sure that the iteration goes all the way to the end?

If so, it might be a good idea to add a comment.

Also, "0" is a bit of an extreme value.

GaelVaroquaux commented on doc/modules/kernel_approximation.rst:
```diff
@@ -59,13 +59,13 @@ a linear algorithm, for example a linear SVM::
     >>> rbf_feature = RBFSampler(gamma=1, random_state=1)
     >>> X_features = rbf_feature.fit_transform(X)
-    >>> clf = SGDClassifier()   # doctest: +NORMALIZE_WHITESPACE
+    >>> clf = SGDClassifier(tol=0.)   # doctest: +NORMALIZE_WHITESPACE
```
Was this really necessary here?

Also, using "tol=0" in an example seems like a bad habit to give to a user. The tol should be set to a reasonable value, not zero. And if people are using SGD, I would expect that they don't care about an excellent convergence, as this is not in general the best algorithm for that.

GaelVaroquaux commented on doc/modules/linear_model.rst:
```diff
@@ -1241,7 +1241,8 @@ This way, we can solve the XOR problem with a linear classifier::
            [1, 1, 0, 0],
            [1, 1, 1, 1]])
-    >>> clf = Perceptron(fit_intercept=False, n_iter=10, shuffle=False).fit(X, y)
+    >>> clf = Perceptron(fit_intercept=False, tol=0, max_iter=10,
```
tol=0 here also? It is necessary in all these places? It worries me a bit.

GaelVaroquaux commented on doc/modules/sgd.rst:
```diff
@@ -59,13 +59,13 @@ for the training samples::
     >>> X = [[0., 0.], [1., 1.]]
     >>> y = [0, 1]
-    >>> clf = SGDClassifier(loss="hinge", penalty="l2")
+    >>> clf = SGDClassifier(loss="hinge", penalty="l2", max_iter=5, tol=0.)
```
tol=0 here too, and in an example :(

GaelVaroquaux reviewed (changes requested):
I haven't reviewed all the changes, because I would first like the choice of specifying tol is so many places to be revisited. If things don't work with the default tol, it's a problem. Are all those tol specification needed?

TomDLT commented:
@GaelVaroquaux
My first reflex had been to make everything work as before in benchmarks and examples, yet I should have reconsidered these changes in a second pass.
I have changed all examples and benchmarks.

amueller commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -68,8 +71,29 @@ def __init__(self, loss, penalty='l2', alpha=0.0001, C=1.0,
+                          "From 0.21, default tol value will be changed to "
+                          "1e-3.", FutureWarning)
+
+        if max_iter == 5:
```
Why 5 and not ``None``? Also, we currently require the user to set both ``tol`` and ``max_iter`` to get rid of the warnings, right?
That's slightly annoying. What we could do but would be slightly magic is "if the user set tol or max_iter, set the other one to a reasonable value (the future default) by default". @GaelVaroquaux @jnothman what do you think?

I like for users to have an easy way to future-proof, warning-free code.

amueller commented on doc/modules/sgd.rst:
```diff
@@ -109,7 +109,7 @@ Using ``loss="log"`` or ``loss="modified_huber"`` enables the
 :math:`P(y|x)` per sample :math:`x`::
 
-    >>> clf = SGDClassifier(loss="log").fit(X, y)
+    >>> clf = SGDClassifier(loss="log", max_iter=5).fit(X, y)
```
why that change?

amueller reviewed (commented):
I'm not sure if we should change any parameters in any of the examples or tests. If they run well with the current default params, then they we can just rename ``n_iter`` to ``max_iter`` and they will still run. It'll produce a ``FutureWarning`` though. Then in 0.21 the default should be good enough to still make it run. Or was adding ``tol`` in some places to ensure that ``1e-3`` actually works sensibly? Then we maybe should also set ``max_iter=100``?

jnothman commented:
I don't mind that solution, @amueller, but it's harder to document and
explain to the user than the current.

On 29 November 2016 at 08:54, Andreas Mueller <notifications@github.com>
wrote:

> *@amueller* commented on this pull request.
>
> I'm not sure if we should change any parameters in any of the examples or
> tests. If they run well with the current default params, then they we can
> just rename n_iter to max_iter and they will still run. It'll produce a
> FutureWarning though. Then in 0.21 the default should be good enough to
> still make it run. Or was adding tol in some places to ensure that 1e-3
> actually works sensibly? Then we maybe should also set max_iter=100?
> ------------------------------
>
> In sklearn/linear_model/stochastic_gradient.py
> <https://github.com/scikit-learn/scikit-learn/pull/5036#pullrequestreview-10421221>
> :
>
> > +        if n_iter is not None:
> +            warnings.warn("n_iter parameter is deprecated in 0.19 and will be"
> +                          " removed in 0.21. Use max_iter and tol instead.",
> +                          DeprecationWarning)
> +            self.max_iter = n_iter
> +            self.tol = 0.
> +
> +        if tol is None:
> +            warnings.warn("tol parameter has been added in 0.19."
> +                          "Set it to a positive value to stop the method "
> +                          "before maximum number of iteration. "
> +                          "Set it to 0 to force the number of iteration. "
> +                          "From 0.21, default tol value will be changed to "
> +                          "1e-3.", FutureWarning)
> +
> +        if max_iter == 5:
>
> Why 5 and not None? Also, we currently require the user to set both tol
> and max_iter to get rid of the warnings, right?
> That's slightly annoying. What we could do but would be slightly magic is
> "if the user set tol or max_iter, set the other one to a reasonable value
> (the future default) by default". @GaelVaroquaux
> <https://github.com/GaelVaroquaux> @jnothman <https://github.com/jnothman>
> what do you think?
>
> I like for users to have an easy way to future-proof, warning-free code.
> ------------------------------
>
> In doc/modules/sgd.rst
> <https://github.com/scikit-learn/scikit-learn/pull/5036#pullrequestreview-10421221>
> :
>
> > @@ -109,7 +109,7 @@ Using ``loss="log"`` or ``loss="modified_huber"`` enables the
>  ``predict_proba`` method, which gives a vector of probability estimates
>  :math:`P(y|x)` per sample :math:`x`::
>
> -    >>> clf = SGDClassifier(loss="log").fit(X, y)
> +    >>> clf = SGDClassifier(loss="log", max_iter=5).fit(X, y)
>
> why that change?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/5036#pullrequestreview-10421221>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6xIg5pAINUl6I6_aWLR5jpXTKO48ks5rC02TgaJpZM4FgXmt>
> .
>

amueller commented:
yeah it's a tradeoff between explicitness and convenience...

TomDLT commented:
**Update**
I have changed the default to `max_iter=None, tol=None, n_iter=None`.
- If nothing is set, it raises a FutureWarning but keep behavior unchanged: `max_iter=5`, and `tol=0`
- If `n_iter` is set, it raises a DeprecationWarning but keep behavior unchanged: `max_iter=n_iter`, and `tol=0`
- If one of `tol` and `max_iter` is set but not the other, the other is set to its future default value (`max_iter=100` and `tol=1e-3`), and no warning is raised (aka **magic**). This results in a future proof code with less warnings.
- If both `tol` and `max_iter` are set, no warning is raised

Note: This behavior in `__init__` breaks [`check_parameters_default_constructible`](https://github.com/scikit-learn/scikit-learn/pull/5036/files#diff-a95fe0e40350c536a5e303e87ac979c4R1534) , so I skipped a part of this check, but I think it's the simplest solution.

Note: I hope that the docstring and warnings are explicit enough about the **magic** change in parameters.

Also, I have simplified at best the examples and benchmarks, and removed all ConvergenceWarning, DeprecationWarning, and FutureWarning from the tests.

amueller commented:
sorry I'm slow. Why does the test break?

TomDLT commented:
No idea. I started a new run.

jnothman commented on sklearn/ensemble/tests/test_base.py:
```diff
@@ -24,8 +24,8 @@
-    ensemble = BaggingClassifier(base_estimator=Perceptron(random_state=None),
-                                 n_estimators=3)
+    ensemble = BaggingClassifier(
+        base_estimator=Perceptron(tol=1e-1, random_state=None), n_estimators=3)
```
If these will work with default tolerance settings before and after the deprecation, I think it would be best to leave it unspecified.

TomDLT replied:
I set it to avoid a `FutureWarning`, and I don't set it to `1e-3` to avoid a `ConvergenceWarning`.
It is quite hard to have a default `tol` that fits for every tests.

jnothman commented:
@tomdlt, could you please fix the merge conflict? Thanks.

TomDLT commented:
Done

jnothman commented:
I'd appreciate if @amueller would confirm that he prefers the new interface as described [above](https://github.com/scikit-learn/scikit-learn/pull/5036#issuecomment-263576418). I think we should aim to merge this soon.

TomDLT commented:
**Update:** New goal for the sprint: be consistent with the sgd solver in MLP classes.

### MLP classes implement:
- validation-based early stopping, computing the score on a validation set (`early_stopping=True, validation_fraction=0.1`)
- convergence-based early stopping, computing the loss on the training set (`early_stopping=False`)

In both case, it stops if the score/loss does not improve twice in a row, with an absolute tolerance `tol`.

### SGD classes implement:
- in master: no stopping, using a fixed number of iteration `n_iter`
- In this PR: convergence-based early stopping, stopping when `max(change in weights) / max(weights) < tol`

This convergence-based early stopping is chosen to be consistent with the one in SAG/SAGA solvers available in `LogisticRegression` and `Ridge`. It is also cheaper than computing the loss on the training set.

**Question:** which stopping criterion should we use when we don't use a validation set? @ogrisel 
Note: I plan to implement validation-based early stopping anyway.

amueller commented:
I have not found the MLP implementation to work well. I think the API is right, but maybe the tolerance etc is not good?

amueller commented on sklearn/ensemble/tests/test_bagging.py:
```diff
@@ -65,7 +65,7 @@ def test_classification():
     for base_estimator in [None,
                            DummyClassifier(),
-                           Perceptron(),
+                           Perceptron(tol=1e0),
```
why? worst case is 5 iterations, right?

amueller replied:
ah, I guess for future? but we get a future warning then, right?

TomDLT replied:
I changed the tests to use the future parameters and avoid a FutureWarning.
I didn't use the future default values to avoid a ConvergenceWarning

amueller commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -67,6 +68,39 @@ def __init__(self, loss, penalty='l2', alpha=0.0001, C=1.0,
+            self.max_iter = 5
+            self.tol = 0.
+
+        elif tol is not None and max_iter is None:
```
why not
```python
else:
   self.tol = tol if tol is not None else 1e-3
   self.max_iter = max_iter if max_iter is not None else 100
 ```

TomDLT replied:
sure

amueller commented:
LGTM apart from nitpick simplification. Have you checked that the tests raise no FutureWarnings?

TomDLT commented:
> Have you checked that the tests raise no FutureWarnings?

Yes, there is not FutureWarning, except in the doctest.
However, I don't like tweaking the docstring example just to avoid a warning.

amueller commented:
Sounds good! Maybe let's wait for @GaelVaroquaux as he requested changes earlier?

TomDLT commented:
## Major update

After discussing with @ogrisel, we decided to be consistent with MLP classes in term of stopping criterion. Instead of stopping when `max(change in weights) / max(weights) < tol`, which can be quite noisy with SGD (contrarily to SAG), we stop when `loss > previous_loss - tol`, as in MLP classes.

Therefore, I updated the default `tol` to `-np.inf` to avoid breaking user's code. I also updated the tests.

Adding a stopping criterion monitoring the score on a validation set is let to a future PR.

jnothman commented:
Test failure at https://travis-ci.org/scikit-learn/scikit-learn/jobs/240037379. I'm cancelling the other travis tests due to backlog.

amueller commented:
@jnothman @lesteve just enabled auto-cancelling, also we all hang out in https://gitter.im/raghavrv

TomDLT committed:
ENH Deprecate n_iter in SGDClassifier and implement max_iter

TomDLT commented:
Note that AppVeyor is already green on my fork:

https://ci.appveyor.com/project/TomDLT/scikit-learn/build/1.0.188

TomDLT committed:
fix common tests

TomDLT committed:
pep8

amueller commented:
What are we waiting for? @GaelVaroquaux ? Appveyor?

TomDLT commented:
I just fixed some conflicts with `estimator_checks.py`.
Let's wait for AppVeyor

TomDLT commented:
How is it even a failure ?

```pytb
AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0
(mismatch 100.0%)
 x: array([  1.000001e+00,  -6.044411e-07,   3.061235e-07,   1.000000e+00,
         9.999993e-01,   1.705852e-06,   1.000005e+00,   7.273812e-07,
         1.000000e+00,   9.999981e-01,  -1.059240e-06,   9.078597e-07,...
 y: array([  1.000001e+00,  -6.044411e-07,   3.061235e-07,   1.000000e+00,
         9.999993e-01,   1.705852e-06,   1.000005e+00,   7.273812e-07,
         1.000000e+00,   9.999981e-01,  -1.059240e-06,   9.078597e-07,...
[success] 5.48% sklearn.decomposition.tests.test_dict_learning.test_sparse_encode_shapes_omp: 13.8399s
```

agramfort commented:
you only see 6 digits after the comma
and you ask for rtol=1e-07 ?

TomDLT commented:
The AppVeyor failures are on master #9111

jnothman commented:
Are we still waiting for @GaelVaroquaux here?

amueller commented:
@TomDLT can you ask @GaelVaroquaux if he wants to review?

ogrisel commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -66,6 +67,29 @@ def __init__(self, loss, penalty='l2', alpha=0.0001, C=1.0,
+                "are left unset, they default to max_iter=5 and tol=-inf. "
+                "If you set one of them, the other one is set to its future "
+                "default value: From 0.21, default max_iter will be 100, "
+                "and default tol will be 1e-3.", FutureWarning)
```
I would put `type(self)` somewhere in the error message. The user is likely to get a warning when running existing code after an upgrade without necessarily remembering that `SGDClassifier / Regressor` is used in his/her code.

ogrisel commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -66,6 +67,29 @@ def __init__(self, loss, penalty='l2', alpha=0.0001, C=1.0,
+                          DeprecationWarning)
+            # Same behavior as before 0.19
+            self.max_iter = n_iter
+            self.tol = -np.inf
```
Also please add a check in `_validate_params` that raises `ValueError` if `self.tol < 0` telling that `tol` should be positive (or zero to disable the convergence criterion).

TomDLT replied:
I copied the behavior of MLP-classes: 
>When the loss or score is not improving by at least tol for two consecutive iterations, convergence is considered to be reached and training stops.

In this case, a negative `tol` can be used (and there is no `ValueError` for negative `tol`).
The `-np.inf` is only here not to change the behavior of default params, but I can use `None` instead if it's less confusing.

TomDLT replied:
But with `tol=None`, it breaks the logic of `FutureWarning`

ogrisel replied:
+1 for None then.

ogrisel reviewed (changes requested):
I have the following comments. In particular I find the use of `-np.inf` and negative  tolerance disturbing. Do you really think negative tolerance is useful feature? If not I would enforce positivity and 0 to disable convergence checks.

TomDLT committed:
change default max_iter to 1000 and improve warnings

TomDLT commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -66,6 +67,29 @@ def __init__(self, loss, penalty='l2', alpha=0.0001, C=1.0,
+            self.tol = -np.inf
+
+        else:
+            self.max_iter = max_iter if max_iter is not None else 100
```
Done

TomDLT committed:
change  tol=-np.inf into tol=None

ogrisel commented on sklearn/linear_model/stochastic_gradient.py:
```diff
@@ -77,17 +77,17 @@ def __init__(self, loss, penalty='l2', alpha=0.0001, C=1.0,
-                "default value: From 0.21, default max_iter will be 100, "
-                "and default tol will be 1e-3.", FutureWarning)
+                "max_iter and tol parameters have been added in %s in 0.19. If"
+                "both are left unset, they default to max_iter=5 and tol=-inf."
```
`tol=None`

ogrisel commented on benchmarks/bench_covertype.py:
```diff
@@ -102,7 +102,7 @@ def load_data(dtype=np.float32, order='C', random_state=13):
     'RandomForest': RandomForestClassifier(n_estimators=20),
     'CART': DecisionTreeClassifier(min_samples_split=5),
-    'SGD': SGDClassifier(alpha=0.001, n_iter=2),
+    'SGD': SGDClassifier(alpha=0.001, max_iter=2, tol=-np.inf),
```
I would rather put the future default: `max_iter=1000` and `tol=1e-3`.

ogrisel commented on benchmarks/bench_sgd_regression.py:
```diff
@@ -70,30 +70,30 @@
-                               n_iter=n_iter, learning_rate="invscaling",
-                               eta0=.01, power_t=0.25)
+                               max_iter=max_iter, learning_rate="invscaling",
+                               eta0=.01, power_t=0.25, tol=-np.inf)
```
Same remark here: just let the convergence check do its work: set `max_iter=1000` and `tol=1e-3`.

ogrisel commented:
It seems that something is broken when I try on this binary version of iris:

```python
from sklearn.datasets import load_iris
from sklearn.preprocessing import scale
from sklearn.linear_model import SGDClassifier


iris = load_iris()
X, y = scale(iris.data), iris.target
y = y != 0


m = SGDClassifier(max_iter=1000, tol=1e-3, verbose=1).fit(X, y)
print('effective n_iter:', m.n_iter_)
```

I get the following output:

```
Editing... done. Executing edited code...
-- Epoch 1
Norm: 25.31, NNZs: 4, Bias: -0.243902, T: 150, Avg. loss: 0.017509
Total training time: 0.00 seconds.
-- Epoch 2
Norm: 22.38, NNZs: 4, Bias: -0.243902, T: 300, Avg. loss: 0.008755
Total training time: 0.00 seconds.
convergence after 2 epochs took 0.00 seconds
effective n_iter: 2
```

I would have expected a larger number of iterations because the loss has decreased by more than `1e-3` during the last iteration.

I tweaked the tolerance to no avail.

ogrisel commented on sklearn/linear_model/tests/test_sgd.py:
```diff
@@ -1158,12 +1165,77 @@ def test_large_regularization():
+    # With tol > -np.inf, the number of iteration should be less than max_iter
+    max_iter = 2000
+    model_1 = SGDClassifier(tol=0, max_iter=max_iter).fit(X, y)
+    assert_greater(max_iter, model_1.n_iter_)
```
Can you please fix the `random_state` of `SGDClassifier` and max sure that `assert_greater(model_1.n_iter_, 5)` for instance?

ogrisel commented on sklearn/linear_model/tests/test_sgd.py:
```diff
@@ -1158,12 +1165,77 @@ def test_large_regularization():
+    # A larger tol should yield a smaller number of iteration
+    model_2 = SGDClassifier(tol=1e-3, max_iter=max_iter).fit(X, y)
+    assert_greater(model_1.n_iter_, model_2.n_iter_)
+
```
Please also add a similar lower bound with a fixed `random_state` (e.g. `n_iter_ > 3` or something).

ogrisel commented on sklearn/linear_model/tests/test_sgd.py:
```diff
@@ -1158,12 +1165,77 @@ def test_large_regularization():
+    # When n_iter=None, and at least one of tol and max_iter is specified
+    assert_no_warnings(init, 100, None, None)
+    assert_no_warnings(init, None, 1e-3, None)
+    assert_no_warnings(init, 100, 1e-3, None)
```
Please add assertions for the resulting values of `clf.max_iter` and `clf.tol` for each of these cases.

TomDLT replied:
cf. `test_tol_and_max_iter_default_values` ?

ogrisel replied:
Fine.

TomDLT committed:
fix sumloss computation

TomDLT commented:
> I would have expected a larger number of iterations because the loss has decreased by more than 1e-3 during the last iteration.

Good point, the loss accumulator was no reset after each epoch, and was not scaled by `n_samples`. I fixed it. Now your snippet gives:
```
-- Epoch 1
Norm: 24.28, NNZs: 4, Bias: 9.764409, T: 150, Avg. loss: 0.222557
Total training time: 0.00 seconds.
-- Epoch 2
Norm: 29.40, NNZs: 4, Bias: 9.535794, T: 300, Avg. loss: 0.195226
Total training time: 0.00 seconds.
-- Epoch 3
Norm: 26.35, NNZs: 4, Bias: 9.535794, T: 450, Avg. loss: 0.000000
Total training time: 0.00 seconds.
-- Epoch 4
Norm: 23.88, NNZs: 4, Bias: 9.535794, T: 600, Avg. loss: 0.000000
Total training time: 0.00 seconds.
convergence after 4 epochs took 0.00 seconds
effective n_iter: 4
```

ogrisel commented:
@TomDLT I have pushed two small improvements to my ogrisel/sgd_maxiter branch. Could you please include them in your PR to get CI to run on them. Other than that I think I am +1.

ogrisel commented:
@TomDLT please allow others scikit-learn devs to push into the branch of your next PRs in the future :)

ogrisel reviewed (approved):
+1 once ogrisel/sgd_maxiter is included to this PR.

TomDLT committed:
merge ogrisel branch

TomDLT commented:
Done and all green

TomDLT committed:
minor fixes

ogrisel commented:
Merged :tada:

jnothman commented:
well done! it's been a long haul!

On 24 Jun 2017 5:49 am, "Olivier Grisel" <notifications@github.com> wrote:

> Merged 🎉
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/5036#issuecomment-310757403>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6zjBIFNNZ8N-hvtFjJh72vFr6JaPks5sHBbigaJpZM4FgXmt>
> .
>

amueller commented:
Wohooo!!
